### PR TITLE
Add support for collecting large files with GRR

### PR DESCRIPTION
**Pull request by tomchop**

### CODE REVIEW

File: dftimewolf/lib/collectors/grr_hosts.py [hunk 1]
```diff
@@ -317,6 +315,56 @@ def _CheckSkippedFlows(self) -> None:
               client_id, flow_id, self.reason
         ))
 
+  def _DownloadArchive(
+    self, grr_flow: Client.Flow, flow_output_dir: str
+  ) -> None:
```
Comment (Onager) [inline dftimewolf/lib/collectors/grr_hosts.py:320]: Please add docstrings here for the args

File: dftimewolf/lib/collectors/grr_hosts.py [hunk 2]
```diff
@@ -317,6 +315,56 @@ def _CheckSkippedFlows(self) -> None:
...
+    output_file_path = os.path.join(self.output_path, f"{grr_flow.flow_id}.zip")
+    file_archive = grr_flow.GetFilesArchive()
+    file_archive.WriteToFile(output_file_path)
+    with zipfile.ZipFile(output_file_path) as archive:
+      archive.extractall(path=flow_output_dir)
+    os.remove(output_file_path)
+
+  def _DownloadBlobs(
+    self, client: Client, pathspecs: Client.Flow, flow_output_dir: str
```
Comment (Onager) [inline dftimewolf/lib/collectors/grr_hosts.py:329]: Please add docstring

File: dftimewolf/lib/collectors/grr_hosts.py [hunk 3]
```diff
@@ -317,6 +315,56 @@ def _CheckSkippedFlows(self) -> None:
...
+      self.logger.debug(f"Downloading blob {filename} from {vfspath}")
+      try:
+        with open(os.path.join(base_dir, filename), "wb") as out:
+          f.GetBlobWithOffset(0).WriteToStream(out)
+      except grr_errors.ResourceNotFoundError as e:
+        self.logger.warning(
+          f"Failed to download blob {filename} from {vfspath}: {e}"
+        )
+
+  def _DownloadTimeline(
```
Comment (Onager) [inline dftimewolf/lib/collectors/grr_hosts.py:350]: Please add docstrings

File: dftimewolf/lib/collectors/grr_hosts.py [hunk 4]
```diff
@@ -317,6 +315,56 @@ def _CheckSkippedFlows(self) -> None:
...
+    ntfs_inodes = client.data.os_info.system.lower() == "windows"
+    timeline = grr_flow.GetCollectedTimelineBody(
+      timestamp_subsecond_precision=True,
+      inode_ntfs_file_reference_format=ntfs_inodes,
+      backslash_escape=True,
+    )
+    timeline.WriteToFile(final_bodyfile_path)
+    return final_bodyfile_path
+
   # TODO: change object to more specific GRR type information.
```
Comment (Onager) [inline dftimewolf/lib/collectors/grr_hosts.py:368]: This looks like it no longer applies, as the type hint here has a more up-to-date definition - while you're here, please update the docstring and remove the todo

File: dftimewolf/lib/collectors/grr_hosts.py (hunk 2 above)
Comment (Onager) [reply dftimewolf/lib/collectors/grr_hosts.py:329]: Type is wrong for "pathspecs" here, should be a list of pathspecs

File: dftimewolf/lib/collectors/grr_hosts.py [hunk 5]
```diff
@@ -329,44 +377,40 @@ def _DownloadFiles(self, client: Client, flow_id: str) -> Optional[str]:
...
-    if not os.path.isdir(client_output_folder):
-      os.makedirs(client_output_folder)
-
-    if is_timeline_flow:
-      shutil.copy2(
-          output_file_path,
-          os.path.join(client_output_folder,
-                       f'{flow_id}_timeline.body'))
+    flow_output_dir = os.path.join(self.output_path, fqdn, flow_id)
+    if not os.path.isdir(flow_output_dir):
```
Comment (Onager) [inline dftimewolf/lib/collectors/grr_hosts.py:383]: This check seems redundant now you're passing exist_ok - remove?

**Review by Onager — CHANGES_REQUESTED** [review]

File: dftimewolf/lib/collectors/grr_hosts.py (hunk 1 above)
Comment (tomchop) [reply dftimewolf/lib/collectors/grr_hosts.py:320]: Done.

**Review by Onager — DISMISSED** [review]

**Comment by tomchop** [comment]
Got an LGTM, merging

**Pull request merged**